k2/csrc/fsa_algo.cc

```diff
@@ -0,0 +1,114 @@
+// k2/csrc/fsa_algo.cc
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_algo.h"
+
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <stack>
+
+namespace {
+
+struct DfsState {
+  int32_t s;
+  int32_t begin;
+  int32_t end;
+};
+
+enum class DfsStateColor : int8_t {
+  kWhite,  // a state which has not been discovered
```

Comment:
For me it might be clearer to just use bool and name it 'discovered' or something like that, but up to you.

Reply:
Fixed.